### PR TITLE
[boost-modular-build-helper] Wrap compiler flags with quotes

### PR DESCRIPTION
**Pull request by ahojnnes**
Without this PR, building boost packages under Windows with custom compiler flags fails as follows:
```
notice: Loading user-config configuration file 'user-config.jam' from '.../vcpkg/buildtrees/boost-container/x64-windows-rel'.
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:6: Unescaped special character in argument <cxxflags>/guard:cf
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:7: Unescaped special character in argument <cflags>/guard:cf
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/machine:x64
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/INCREMENTAL:NO
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/OPT:REF
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/OPT:ICF
...\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/guard:cf
```

### CODE REVIEW

File: ports/boost-modular-build-helper/CMakeLists.txt [hunk 1]
```diff
@@ -122,41 +122,42 @@ string(REGEX REPLACE "[ \t\r\n]+" " " LDFLAGS "${LDFLAGS}")
...
-    string(REPLACE " " " <cxxflags>" CXXFLAGS "<cxxflags>${CXXFLAGS}")
+    string(REPLACE " " "\" <cxxflags>\"" CXXFLAGS "<cxxflags>\"${CXXFLAGS}\"")
 endif()
 if(NOT CFLAGS STREQUAL "")
-    string(REPLACE " " " <cflags>" CFLAGS "<cflags>${CFLAGS}")
+    string(REPLACE " " "\" <cflags>\"" CFLAGS "<cflags>\"${CFLAGS}\"")
 endif()
 if(NOT LDFLAGS STREQUAL "")
-    string(REPLACE " " " <linkflags>" LDFLAGS "<linkflags>${LDFLAGS}")
+    list(APPEND B2_OPTIONS linkflags="${LDFLAGS}")
```
Comment (ahojnnes) [inline ports/boost-modular-build-helper/CMakeLists.txt:131]: Apparently, there is either something wrong with how linkflags are currently specified in the user config or there is a bug in b2e... Without this line here, the linker flags are not correctly applied to the linker commands during execution under Windows.

Comment (ahojnnes) [reply ports/boost-modular-build-helper/CMakeLists.txt:131]: Anybody an idea what's going on?

Comment (dg0yt) [reply ports/boost-modular-build-helper/CMakeLists.txt:131]: I don't know much about the boost build system and this specific issue, but...
~~~
 list(APPEND B2_OPTIONS linkflags="${LDFLAGS}")
~~~
is shell quoting applied in CMake, with the proper portable syntax being:
~~~
 list(APPEND B2_OPTIONS "linkflags=${LDFLAGS}")
~~~

**Comment by PhoebeHui** [comment]
@ahojnnes, could you report an issue for this PR? what compiler flag do you apply?

**Comment by ahojnnes** [comment]
I am trying to be a good citizen :-) and enable /guard:cf and other related flags to address binskim suggestions.

**Comment by PhoebeHui** [comment]
@ahojnnes, I reproed the issue when I added the flag to triplets\x64-window.cmake, and test boost-container.

```
set(VCPKG_CXX_FLAGS "/guard:cf")
set(VCPKG_C_FLAGS "/guard:cf")
```
cc @yurybura

**Comment by ahojnnes** [comment]
@PhoebeHui Yes, exactly, that's what I was trying to achieve.

**Comment by yurybura** [comment]
@PhoebeHui @ahojnnes If I understand correctly build doesn't fail.
Boost.Build reports warning but applies corresponding flags.
My log:
```
[...]
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:6: Unescaped special character in argument <cxxflags>/guard:cf
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:7: Unescaped special character in argument <cflags>/guard:cf
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/machine:x64
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/INCREMENTAL:NO
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/OPT:REF
D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\user-config.jam:8: Unescaped special character in argument <linkflags>/OPT:ICF
[...]
    call "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\standalone\msvc\19858a0c092d6daed3b7579665d68f18\msvc-setup.bat" amd64 >nul
 cl "..\src\dlmalloc.cpp" -c -Fo"D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\dlmalloc.obj"     -TP /wd4675 /EHs /GR /Zc:throwingNew /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP /guard:cf /MD /O2 /Oi /Gy /DNDEBUG /Z7 /O2 /Z7 /Ob2 /W3 /MD /Zc:forScope /Zc:wchar_t /Zc:inline /Gw /favor:blend -DBOOST_ALL_NO_LIB=1 -DBOOST_CONTAINER_DYN_LINK=1 -DNDEBUG "-I..\include" "-ID:\libs\vcpkg\installed\x64-windows\include" 
[...]
        call "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\standalone\msvc\19858a0c092d6daed3b7579665d68f18\msvc-setup.bat" amd64 >nul
 link /NOLOGO /INCREMENTAL:NO "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\alloc_lib.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\dlmalloc.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\global_resource.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\monotonic_buffer_resource.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\pool_resource.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\synchronized_pool_resource.obj" "D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\unsynchronized_pool_resource.obj"      /DEBUG /MACHINE:X64 /MANIFEST:EMBED /OPT:REF,ICF /subsystem:console /out:"D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\boost_container-vc142-mt-x64-1_78.dll"   /DLL /IMPLIB:"D:\libs\vcpkg\buildtrees\boost-container\x64-windows-rel\boost\build\aa84e6e880a9d383ddf248bc50c22218\boost_container-vc142-mt-x64-1_78.lib" 
[...]
```

**Comment by ahojnnes** [comment]
@yurybura The warnings are not a blocker, but see the annotated code line above. Without this, the linker flags are just ignored.

**Comment by yurybura** [comment]
> @yurybura The warnings are not a blocker, but see the annotated code line above. Without this, the linker flags are just ignored.

@ahojnnes Ok, thank you. I think this is the real issue - Boost.Build ignores linker flags defined for msvc toolset in user_config.jam.

**Comment by yurybura** [comment]
> > @yurybura The warnings are not a blocker, but see the annotated code line above. Without this, the linker flags are just ignored.
> 
> @ahojnnes Ok, thank you. I think this is the real issue - Boost.Build ignores linker flags defined for msvc toolset in user_config.jam.

@Kojoley Could you help to understand why linker flags defined in the user_config.jam are ignored? Thanks in advance.

**Comment by Kojoley** [comment]
> @Kojoley Could you help to understand why linker flags defined in the user_config.jam are ignored? Thanks in advance.

That's a bug in B2, and it seems that there should be the same bug for asmflags. For some historical reasons msvc.jam uses different names than common.handle-options and other toolsets use.

**Comment by PhoebeHui** [comment]
Relate to https://github.com/microsoft/vcpkg/pull/20697

**Review by PhoebeHui — APPROVED** [review]

**Comment by ras0219-msft** [comment]
Adding this doc link for posterity: https://grafikrobot.github.io/b2doc/#jam.language.lexical

**Comment by ras0219-msft** [comment]
Looking at the documentation, it seems like we can actually take a much simpler approach here.

```
using gcc : 5 : : <cxxflags>"-std=c++14 -O2" ;
```

is claimed to work. This means that instead of trying to split on spaces, we should be able to do a very simple fix of:
1. Get everything concatenated into a single string a la CXXFLAGS/CFLAGS like all the other buildsystems
2. Escape `\` and `"` with `\` (either as one step via regex or two simple string replacements)
3. Wrap the whole enchilada in quotes and prepend the appropriate `<fooflags>`.

I believe this should fix multiple current issues, including this one and https://github.com/microsoft/vcpkg/issues/22840.

I think when I originally wrote this code I didn't find that part of the docs and assumed that boost build would apply its own layers of escaping to the arguments (such as surrounding every `<cxxflags>` entry in quotes to ensure it was passed as a single argument to the compiler). That's why we didn't previously take the approach I've described above.

**Comment by PhoebeHui** [comment]
Ping @ahojnnes for your response!

**Comment by ahojnnes** [comment]
What @ras0219-msft proposes seems like the right approach and simplifies the whole script. I am supporting to take this approach, but I don't have any cycles to tackle the issues anytime soon to be honest.

**Comment by PhoebeHui** [comment]
Closing in favor of https://github.com/microsoft/vcpkg/pull/23001.

**Pull request closed without merging**